<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Marriages-Worldwide" data-toc-modified-id="Marriages-Worldwide-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Marriages Worldwide</a></span></li><li><span><a href="#Extramarital-Affairs" data-toc-modified-id="Extramarital-Affairs-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Extramarital Affairs</a></span></li></ul></div>

## Marriages Worldwide

https://en.wikipedia.org/wiki/Divorce_demography Divorce statistics by country/region (per 1,000 population / year)

https://en.wikipedia.org/wiki/List_of_countries_by_alcohol_consumption_per_capita

https://en.wikipedia.org/wiki/Importance_of_religion_by_country

In [23]:
marriages <- read.table("marriages.csv", sep = ",", header = T)
head(marriages)

Country               Continent MarriageRate DivorceRate Ratio Percent
1 Albania               Europe    8.1          1.7         4.76  20.99  
2 Algeria               Africa    10.1         1.6         6.31  15.84  
3 Armenia               Europe    6.0          1.0         6.00  16.67  
4 Australia (more info) Oceania   4.6          2.0         2.30  43.48  
5 Austria               Europe    5.3          1.8         2.94  33.96  
6 Azerbaijan            Asia      9.7          1.5         6.47  15.46  
  DataSourceYear
1 (2018)[1]     
2 (2013)[2]     
3 (2011)[3][4]  
4 (2017)[5]     
5 (2018)[1]     
6 (2011)[3][4]

In [24]:
marriages$DataSourceYear <- NULL
head(marriages)

Country               Continent MarriageRate DivorceRate Ratio Percent
1 Albania               Europe    8.1          1.7         4.76  20.99  
2 Algeria               Africa    10.1         1.6         6.31  15.84  
3 Armenia               Europe    6.0          1.0         6.00  16.67  
4 Australia (more info) Oceania   4.6          2.0         2.30  43.48  
5 Austria               Europe    5.3          1.8         2.94  33.96  
6 Azerbaijan            Asia      9.7          1.5         6.47  15.46

In [25]:
for (col in marriages) {
    print(mode(col))
}

[1] "numeric"
[1] "numeric"
[1] "numeric"
[1] "numeric"
[1] "numeric"
[1] "numeric"


In [26]:
for (col in marriages) {
    print(class(col))
}

[1] "factor"
[1] "factor"
[1] "factor"
[1] "factor"
[1] "numeric"
[1] "numeric"


In [27]:
as.numeric(marriages$MarriageRate)

[1] 54  3 38 26 33 59 39 57 20  5 28 43 21 24 15 49 11 33 29 32 45 31 35 24 35
 [26]  7 17 30 24 23 18 46 29 44 30 27 19 32 29  8 26 42 14 52 28  4 56 32 59 45
 [51] 58  6 30 47 13 41 37 55 32 50 16 33 18 28 25 42 23 18 31 16 15 51 33 12 37
 [76] 39 35 32 10 45 36 17 17 27 17  2 22 30 28  5 36  9 34 48 40  1 45 44 12 24
[101] 42 14 53 15 36

In [28]:
as.vector(marriages$MarriageRate)

[1] "8.1"  "10.1" "6.0"  "4.6"  "5.3"  "9.7"  "6.1"  "9.2"  "3.9"  "10.6"
 [11] "4.8"  "6.6"  "4.1"  "4.4"  "3.3"  "7.2"  "2.3"  "5.3"  "4.9"  "5.2" 
 [21] "6.8"  "5.1"  "5.6"  "4.4"  "5.6"  "11.0" "3.5"  "5.0"  "4.4"  "4.3" 
 [31] "3.7"  "6.9"  "4.9"  "6.7"  "5.0"  "4.7"  "3.8"  "5.2"  "4.9"  "11.2"
 [41] "4.6"  "6.5"  "3.2"  "7.5"  "4.8"  "10.2" "8.6"  "5.2"  "9.7"  "6.8" 
 [51] "9.5"  "10.8" "5.0"  "7.0"  "3.1"  "6.4"  "5.8"  "8.2"  "5.2"  "7.3" 
 [61] "3.4"  "5.3"  "3.7"  "4.8"  "4.5"  "6.5"  "4.3"  "3.7"  "5.1"  "3.4" 
 [71] "3.3"  "7.4"  "5.3"  "2.8"  "5.8"  "6.1"  "5.6"  "5.2"  "17.4" "6.8" 
 [81] "5.7"  "3.5"  "3.5"  "4.7"  "3.5"  "?"    "4.2"  "5.0"  "4.8"  "10.6"
 [91] "5.7"  "13.5" "5.5"  "7.1"  "6.3"  ""     "6.8"  "6.7"  "2.8"  "4.4" 
[101] "6.5"  "3.2"  "7.8"  "3.3"  "5.7"

In [29]:
as.numeric(as.vector(marriages$MarriageRate))

Warning message in eval(expr, envir, enclos):
“NAs introduced by coercion”

[1]  8.1 10.1  6.0  4.6  5.3  9.7  6.1  9.2  3.9 10.6  4.8  6.6  4.1  4.4  3.3
 [16]  7.2  2.3  5.3  4.9  5.2  6.8  5.1  5.6  4.4  5.6 11.0  3.5  5.0  4.4  4.3
 [31]  3.7  6.9  4.9  6.7  5.0  4.7  3.8  5.2  4.9 11.2  4.6  6.5  3.2  7.5  4.8
 [46] 10.2  8.6  5.2  9.7  6.8  9.5 10.8  5.0  7.0  3.1  6.4  5.8  8.2  5.2  7.3
 [61]  3.4  5.3  3.7  4.8  4.5  6.5  4.3  3.7  5.1  3.4  3.3  7.4  5.3  2.8  5.8
 [76]  6.1  5.6  5.2 17.4  6.8  5.7  3.5  3.5  4.7  3.5   NA  4.2  5.0  4.8 10.6
 [91]  5.7 13.5  5.5  7.1  6.3   NA  6.8  6.7  2.8  4.4  6.5  3.2  7.8  3.3  5.7

In [30]:
marriages$MarriageRate <- as.numeric(as.vector(marriages$MarriageRate))

Warning message in eval(expr, envir, enclos):
“NAs introduced by coercion”

In [31]:
lapply(marriages, class)

$Country
[1] "factor"

$Continent
[1] "factor"

$MarriageRate
[1] "numeric"

$DivorceRate
[1] "factor"

$Ratio
[1] "numeric"

$Percent
[1] "numeric"

In [32]:
apply(marriages, 2, class)

Country    Continent MarriageRate  DivorceRate        Ratio      Percent 
 "character"  "character"  "character"  "character"  "character"  "character"

In [33]:
?apply

Z dokumentace:

If X is not an array but an object of a class with a non-null dim value (such as a data frame), apply attempts to coerce it to an array via as.matrix if it is two-dimensional (e.g., a data frame) or via as.array. 

In [34]:
sapply(marriages, class)

Country    Continent MarriageRate  DivorceRate        Ratio      Percent 
    "factor"     "factor"    "numeric"     "factor"    "numeric"    "numeric"

In [35]:
?sapply

Z dokumentace:

**sapply is a user-friendly version and wrapper of `lapply`** by default returning a vector, matrix or, if simplify = "array", an array if appropriate, by applying simplify2array(). 

In [36]:
marriages$DivorceRate <- as.numeric(as.vector(marriages$DivorceRate))
sapply(marriages, class)

Warning message in eval(expr, envir, enclos):
“NAs introduced by coercion”

Country    Continent MarriageRate  DivorceRate        Ratio      Percent 
    "factor"     "factor"    "numeric"    "numeric"    "numeric"    "numeric"

In [37]:
head(marriages)

Country               Continent MarriageRate DivorceRate Ratio Percent
1 Albania               Europe     8.1         1.7         4.76  20.99  
2 Algeria               Africa    10.1         1.6         6.31  15.84  
3 Armenia               Europe     6.0         1.0         6.00  16.67  
4 Australia (more info) Oceania    4.6         2.0         2.30  43.48  
5 Austria               Europe     5.3         1.8         2.94  33.96  
6 Azerbaijan            Asia       9.7         1.5         6.47  15.46

In [38]:
marriages$MarriageRate / marriages$DivorceRate
100 * marriages$DivorceRate / marriages$MarriageRate

[1]  4.764706  6.312500  6.000000  2.300000  2.944444  6.466667  6.100000
  [8]  2.243902  1.857143  3.925926  8.000000  4.714286  2.733333  2.095238
 [15]  4.714286  2.250000  3.285714  2.120000  3.266667  1.793103  3.090909
 [22]  2.217391  2.153846  2.444444  5.090909  5.789474  4.375000  2.083333
 [29]  2.200000  1.791667  1.947368  5.307692  2.578947  2.093750  4.545455
 [36]  2.611111  3.166667  3.058824  2.722222  7.000000  6.571429  3.611111
 [43]  2.133333  5.000000  2.823529  3.923077  3.739130  2.363636  6.062500
 [50]  2.193548  5.937500  4.320000  2.083333  2.258065  1.550000  4.000000
 [57]  8.285714  5.857143  5.777778  2.354839  3.090909  3.785714  2.055556
 [64]  2.400000  5.625000  8.125000  2.263158  3.700000  3.000000  1.700000
 [71]  3.000000  4.625000  1.358974  4.000000  7.250000  2.440000  2.666667
 [78]  3.714286  9.157895  3.578947  3.166667  3.181818  5.833333  2.136364
 [85]  1.750000        NA  3.230769  2.000000  2.400000  8.153846  2.478261
 [92]  9.642857  3.928571  5.916667  2.863636        NA  4.000000  2.392857
 [99]  4.000000  2.444444  2.241379        NA  7.090909  2.750000 14.250000

[1] 20.987654 15.841584 16.666667 43.478261 33.962264 15.463918 16.393443
  [8] 44.565217 53.846154 25.471698 12.500000 21.212121 36.585366 47.727273
 [15] 21.212121 44.444444 30.434783 47.169811 30.612245 55.769231 32.352941
 [22] 45.098039 46.428571 40.909091 19.642857 17.272727 22.857143 48.000000
 [29] 45.454545 55.813953 51.351351 18.840580 38.775510 47.761194 22.000000
 [36] 38.297872 31.578947 32.692308 36.734694 14.285714 15.217391 27.692308
 [43] 46.875000 20.000000 35.416667 25.490196 26.744186 42.307692 16.494845
 [50] 45.588235 16.842105 23.148148 48.000000 44.285714 64.516129 25.000000
 [57] 12.068966 17.073171 17.307692 42.465753 32.352941 26.415094 48.648649
 [64] 41.666667 17.777778 12.307692 44.186047 27.027027 33.333333 58.823529
 [71] 33.333333 21.621622 73.584906 25.000000 13.793103 40.983607 37.500000
 [78] 26.923077 10.919540 27.941176 31.578947 31.428571 17.142857 46.808511
 [85] 57.142857        NA 30.952381 50.000000 41.666667 12.264151 40.350877
 [92] 10.370370 25.454545 16.901408 34.920635        NA 25.000000 41.791045
 [99] 25.000000 40.909091 44.615385        NA 14.102564 36.363636  7.017544

In [39]:
levels(marriages$Continent)

[1] ""              "Africa"        "Asia"          "Europe"       
[5] "North America" "Oceania"       "South America"

In [40]:
sum(marriages$Continent == '')

[1] 3

In [41]:
marriages[marriages$Continent == '',]

Country   Continent MarriageRate DivorceRate Ratio Percent
10 Bermuda             10.6         2.7         3.93  25.47  
29 EU                   4.4         2.0         2.20  45.45  
34 Gibraltar            6.7         3.2         2.09  47.76

In [42]:
tapply(
    marriages$Country, 
    marriages$Continent, 
    function(x) paste(x, collapse = ", ")
)

"Bermuda, EU, Gibraltar" 
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Africa 
                                                                                                                                                                                                                                                                                                                                                                                                                           "Algeria, Egypt, Libya, Mauritius, Seychelles, South Africa (more info), Tunisia" 
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Asia 
                                                                                                                                                                                                              "Azerbaijan, China, Cyprus, Georgia, Iran, Israel, Japan (more info), Jordan, Kazakhstan, Kuwait, Kyrgyzstan, Lebanon, Malaysia, Mongolia, Qatar, Saudi Arabia, Singapore, South Korea (more info), Sri Lanka, Syria, Taiwan, Tajikistan, Thailand, United Arab Emirates, Uzbekistan, Vietnam" 
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      Europe 
"Albania, Armenia, Austria, Belarus, Belgium (more info), Bosnia and Herzegovina, Bulgaria, Croatia, Czech Republic, Denmark, Estonia, Finland, France, Germany, Greece, Hungary, Iceland, Ireland, Italy, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta (more info), Moldova, Montenegro, Netherlands, Macedonia, Norway (more info), Poland (more info), Portugal, Romania, Russia, San Marino, Serbia, Slovakia, Slovenia, Spain, Sweden, Switzerland, Turkey, Ukraine, United Kingdom (more info)" 
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               North America 
                                                                                                                                                                                                                                                                     "Bahamas, Canada (more info), Costa Rica, Cuba, Dominican Republic, El Salvador, Grenada, Guatemala, Jamaica, Mexico, Nicaragua, Panama, Saint Lucia, Saint Vincent and the Grenadines, Trinidad and Tobago, United States (more info)" 
                      

In [43]:
by(
    marriages, 
    marriages$Continent,
    function(x) {
        mean(x$Percent)
    }
)

marriages$Continent: 
[1] 39.56
------------------------------------------------------------ 
marriages$Continent: Africa
[1] NA
------------------------------------------------------------ 
marriages$Continent: Asia
[1] NA
------------------------------------------------------------ 
marriages$Continent: Europe
[1] 38.79791
------------------------------------------------------------ 
marriages$Continent: North America
[1] 30.30375
------------------------------------------------------------ 
marriages$Continent: Oceania
[1] 34.01667
------------------------------------------------------------ 
marriages$Continent: South America
[1] NA

In [44]:
marriages$Percent

[1] 20.99 15.84 16.67 43.48 33.96 15.46 16.39 44.57 53.85 25.47 12.50 21.21
 [13] 36.59 47.73 21.21 44.44 30.43 47.17 30.61 55.77 32.35 45.10 46.43 40.91
 [25] 19.64 17.27 22.86 48.00 45.45 55.81 51.35 18.84 38.78 47.76 22.00 38.30
 [37] 31.58 32.69 36.73 14.29 15.22 27.69 46.88 20.00 35.42 25.49 26.74 42.31
 [49] 16.49 45.59 16.84 23.15 48.00 44.29 64.52 25.00 12.07 17.07 17.31 42.47
 [61] 32.35 26.42 48.65 41.67 17.78 12.31 44.19 27.03 33.33 58.82 33.33 21.62
 [73] 73.58 25.00 13.79 40.98 37.50 26.92 10.92 27.94 31.58 31.43 17.14 46.81
 [85] 57.14    NA 30.95 50.00 41.67 12.26 40.35 10.37 25.45 16.90 34.92    NA
 [97] 25.00 41.79 25.00 40.91 44.62    NA 14.10 36.36  7.02

In [45]:
na.omit(marriages$Percent)

[1] 20.99 15.84 16.67 43.48 33.96 15.46 16.39 44.57 53.85 25.47 12.50 21.21
 [13] 36.59 47.73 21.21 44.44 30.43 47.17 30.61 55.77 32.35 45.10 46.43 40.91
 [25] 19.64 17.27 22.86 48.00 45.45 55.81 51.35 18.84 38.78 47.76 22.00 38.30
 [37] 31.58 32.69 36.73 14.29 15.22 27.69 46.88 20.00 35.42 25.49 26.74 42.31
 [49] 16.49 45.59 16.84 23.15 48.00 44.29 64.52 25.00 12.07 17.07 17.31 42.47
 [61] 32.35 26.42 48.65 41.67 17.78 12.31 44.19 27.03 33.33 58.82 33.33 21.62
 [73] 73.58 25.00 13.79 40.98 37.50 26.92 10.92 27.94 31.58 31.43 17.14 46.81
 [85] 57.14 30.95 50.00 41.67 12.26 40.35 10.37 25.45 16.90 34.92 25.00 41.79
 [97] 25.00 40.91 44.62 14.10 36.36  7.02
attr(,"na.action")
[1]  86  96 102
attr(,"class")
[1] "omit"

In [46]:
na.omit(marriages)

Country                          Continent     MarriageRate DivorceRate
1   Albania                          Europe         8.1         1.7        
2   Algeria                          Africa        10.1         1.6        
3   Armenia                          Europe         6.0         1.0        
4   Australia (more info)            Oceania        4.6         2.0        
5   Austria                          Europe         5.3         1.8        
6   Azerbaijan                       Asia           9.7         1.5        
7   Bahamas                          North America  6.1         1.0        
8   Belarus                          Europe         9.2         4.1        
9   Belgium (more info)              Europe         3.9         2.1        
10  Bermuda                                        10.6         2.7        
11  Bosnia and Herzegovina           Europe         4.8         0.6        
12  Brazil                           South America  6.6         1.4        
13  Bulgaria                         Europe         4.1         1.5        
14  Canada (more info)               North America  4.4         2.1        
15  Chile                            South America  3.3         0.7        
16  China                            Asia           7.2         3.2        
17  Colombia                         South America  2.3         0.7        
18  Costa Rica                       North America  5.3         2.5        
19  Croatia                          Europe         4.9         1.5        
20  Cuba                             North America  5.2         2.9        
21  Cyprus                           Asia           6.8         2.2        
22  Czech Republic                   Europe         5.1         2.3        
23  Denmark                          Europe         5.6         2.6        
24  Dominican Republic               North America  4.4         1.8        
25  Ecuador                          South America  5.6         1.1        
26  Egypt                            Africa        11.0         1.9        
27  El Salvador                      North America  3.5         0.8        
28  Estonia                          Europe         5.0         2.4        
29  EU                                              4.4         2.0        
30  Finland                          Europe         4.3         2.4        
... ...                              ...           ...          ...        
73  Russia                           Europe         5.3         3.9        
74  Saint Lucia                      North America  2.8         0.7        
75  Saint Vincent and the Grenadines North America  5.8         0.8        
76  San Marino                       Europe         6.1         2.5        
77  Saudi Arabia                     Asia           5.6         2.1        
78  Serbia                           Europe         5.2         1.4        
79  Seychelles                       Africa        17.4         1.9        
80  Singapore                        Asia           6.8         1.9        
81  Slovakia                         Europe         5.7         1.8        
82  Slovenia                         Europe         3.5         1.1        
83  South Africa (more info)         Africa         3.5         0.6        
84  South Korea (more info)          Asia           4.7         2.2        
85  Spain                            Europe         3.5         2.0        
87  Suriname                         South America  4.2         1.3        
88  Sweden                           Europe         5.0         2.5        
89  Switzerland                      Europe         4.8         2.0        
90  Syria                            Asia          10.6         1.3        
91  Taiwan                           Asia           5.7         2.3        
92  Tajikistan                       Asia          13.5         1.4        
93  Thailand                         Asia           5.5         1.4        
94  Tonga                            Oceania        

In [47]:
sum(is.na(na.omit(marriages)))

[1] 0

In [48]:
marriages <- na.omit(marriages)

In [49]:
by(
    marriages, 
    marriages$Continent,
    function(x) {
        mean(x$Percent)
    }
)

marriages$Continent: 
[1] 39.56
------------------------------------------------------------ 
marriages$Continent: Africa
[1] 16.89833
------------------------------------------------------------ 
marriages$Continent: Asia
[1] 26.1536
------------------------------------------------------------ 
marriages$Continent: Europe
[1] 38.79791
------------------------------------------------------------ 
marriages$Continent: North America
[1] 30.30375
------------------------------------------------------------ 
marriages$Continent: Oceania
[1] 34.01667
------------------------------------------------------------ 
marriages$Continent: South America
[1] 26.63333

In [50]:
continents_summary <- by(
    marriages, 
    marriages$Continent, # jestli chceme rozdelit podle vic sloupcu, posleme list
    function(x) {
        mean(x$Percent)
    }
)

continents_summary

marriages$Continent: 
[1] 39.56
------------------------------------------------------------ 
marriages$Continent: Africa
[1] 16.89833
------------------------------------------------------------ 
marriages$Continent: Asia
[1] 26.1536
------------------------------------------------------------ 
marriages$Continent: Europe
[1] 38.79791
------------------------------------------------------------ 
marriages$Continent: North America
[1] 30.30375
------------------------------------------------------------ 
marriages$Continent: Oceania
[1] 34.01667
------------------------------------------------------------ 
marriages$Continent: South America
[1] 26.63333

In [51]:
class(continents_summary)

[1] "by"

In [52]:
continents_summary[['Europe']]

[1] 38.79791

In [53]:
sort(continents_summary)

marriages$Continent
       Africa          Asia South America North America       Oceania 
     16.89833      26.15360      26.63333      30.30375      34.01667 
       Europe               
     38.79791      39.56000

In [54]:
sort(continents_summary, decreasing = T)

marriages$Continent
                     Europe       Oceania North America South America 
     39.56000      38.79791      34.01667      30.30375      26.63333 
         Asia        Africa 
     26.15360      16.89833

In [55]:
split(marriages, marriages$Continent)

[[1]]
     Country Continent MarriageRate DivorceRate Ratio Percent
10   Bermuda                   10.6         2.7  3.93   25.47
29        EU                    4.4         2.0  2.20   45.45
34 Gibraltar                    6.7         3.2  2.09   47.76

$Africa
                    Country Continent MarriageRate DivorceRate Ratio Percent
2                   Algeria    Africa         10.1         1.6  6.31   15.84
26                    Egypt    Africa         11.0         1.9  5.79   17.27
52                    Libya    Africa         10.8         2.5  4.32   23.15
58                Mauritius    Africa          8.2         1.4  5.86   17.07
79               Seychelles    Africa         17.4         1.9  9.16   10.92
83 South Africa (more info)    Africa          3.5         0.6  5.83   17.14

$Asia
                    Country Continent MarriageRate DivorceRate Ratio Percent
6                Azerbaijan      Asia          9.7         1.5  6.47   15.46
16                    China      Asia          7.2         3.2  2.25   44.44
21                   Cyprus      Asia          6.8         2.2  3.09   32.35
32                  Georgia      Asia          6.9         1.3  5.31   18.84
40                     Iran      Asia         11.2         1.6  7.00   14.29
42                   Israel      Asia          6.5         1.8  3.61   27.69
45        Japan (more info)      Asia          4.8         1.7  2.82   35.42
46                   Jordan      Asia         10.2         2.6  3.92   25.49
47               Kazakhstan      Asia          8.6         2.3  3.74   26.74
48                   Kuwait      Asia          5.2         2.2  2.36   42.31
49               Kyrgyzstan      Asia          9.7         1.6  6.06   16.49
51                  Lebanon      Asia          9.5         1.6  5.94   16.84
56                 Malaysia      Asia          6.4         1.6  4.00   25.00
61                 Mongolia      Asia          3.4         1.1  3.09   32.35
71                    Qatar      Asia          3.3         1.1  3.00   33.33
77             Saudi Arabia      Asia          5.6         2.1  2.67   37.50
80                Singapore      Asia          6.8         1.9  3.58   27.94
84  South Korea (more info)      Asia          4.7         2.2  2.14   46.81
90                    Syria      Asia         10.6         1.3  8.15   12.26
91                   Taiwan      Asia          5.7         2.3  2.48   40.35
92               Tajikistan      Asia         13.5         1.4  9.64   10.37
93                 Thailand      Asia          5.5         1.4  3.93   25.45
99     United Arab Emirates      Asia          2.8         0.7  4.00   25.00
103              Uzbekistan      Asia          7.8         1.1  7.09   14.10
105                 Vietnam      Asia          5.7         0.4 14.25    7.02

$Europe
                       Country Continent MarriageRate DivorceRate Ratio Percent
1                      Albania    Europe          8.1         1.7  4.76   20.99
3                      Armenia    Europe          6.0         1.0  6.00   16.67
5                      Austria    Europe          5.3         1.8  2.94   33.96
8                      Belarus    Europe          9.2         4.1  2.24   44.57
9          Belgium (more info)    Europe          3.9         2.1  1.86   53.85
11      Bosnia and Herzegovina    Europe          4.8         0.6  8.00   12.50
13                    Bulgaria    Europe          4.1         1.5  2.73   36.59
19                     Croatia    Europe          4.9         1.5  3.27   30.61
22              Czech Republic    Europe          5.1         2.3  2.22   45.10
23                     Denmark    Europe          5.6         2.6  2.15   46.43
28                     Estonia    Europe          5.0         2.4  2.08   48.00
30                     Finland    Europe          4.3         2.4  1.79   55.81
31                      France    Europe          3.7         1.9  1.95   51.35
33                     Germany    Europe          4.9       

In [56]:
split(marriages, marriages$Continent)$Europe

Country                    Continent MarriageRate DivorceRate Ratio Percent
1   Albania                    Europe    8.1          1.7         4.76  20.99  
3   Armenia                    Europe    6.0          1.0         6.00  16.67  
5   Austria                    Europe    5.3          1.8         2.94  33.96  
8   Belarus                    Europe    9.2          4.1         2.24  44.57  
9   Belgium (more info)        Europe    3.9          2.1         1.86  53.85  
11  Bosnia and Herzegovina     Europe    4.8          0.6         8.00  12.50  
13  Bulgaria                   Europe    4.1          1.5         2.73  36.59  
19  Croatia                    Europe    4.9          1.5         3.27  30.61  
22  Czech Republic             Europe    5.1          2.3         2.22  45.10  
23  Denmark                    Europe    5.6          2.6         2.15  46.43  
28  Estonia                    Europe    5.0          2.4         2.08  48.00  
30  Finland                    Europe    4.3          2.4         1.79  55.81  
31  France                     Europe    3.7          1.9         1.95  51.35  
33  Germany                    Europe    4.9          1.9         2.58  38.78  
36  Greece                     Europe    4.7          1.8         2.61  38.30  
38  Hungary                    Europe    5.2          1.7         3.06  32.69  
39  Iceland                    Europe    4.9          1.8         2.72  36.73  
41  Ireland                    Europe    4.6          0.7         6.57  15.22  
43  Italy                      Europe    3.2          1.5         2.13  46.88  
50  Latvia                     Europe    6.8          3.1         2.19  45.59  
53  Liechtenstein              Europe    5.0          2.4         2.08  48.00  
54  Lithuania                  Europe    7.0          3.1         2.26  44.29  
55  Luxembourg                 Europe    3.1          2.0         1.55  64.52  
57  Malta (more info)          Europe    5.8          0.7         8.29  12.07  
60  Moldova                    Europe    7.3          3.1         2.35  42.47  
62  Montenegro                 Europe    5.3          1.4         3.79  26.42  
63  Netherlands                Europe    3.7          1.8         2.06  48.65  
66  Macedonia                  Europe    6.5          0.8         8.13  12.31  
67  Norway (more info)         Europe    4.3          1.9         2.26  44.19  
69  Poland (more info)         Europe    5.1          1.7         3.00  33.33  
70  Portugal                   Europe    3.4          2.0         1.70  58.82  
72  Romania                    Europe    7.4          1.6         4.63  21.62  
73  Russia                     Europe    5.3          3.9         1.36  73.58  
76  San Marino                 Europe    6.1          2.5         2.44  40.98  
78  Serbia                     Europe    5.2          1.4         3.71  26.92  
81  Slovakia                   Europe    5.7          1.8         3.17  31.58  
82  Slovenia                   Europe    3.5          1.1         3.18  31.43  
85  Spain                      Europe    3.5          2.0         1.75  57.14  
88  Sweden                     Europe    5.0          2.5         2.00  50.00  
89  Switzerland                Europe    4.8          2.0         2.40  41.67  
97  Turkey                     Europe    6.8          1.7         4.00  25.00  
98  Ukraine                    Europe    6.7          2.8         2.39  41.79  
100 United Kingdom (more info) Europe    4.4          1.8         2.44  40.91

In [57]:
aggregate(marriages, list(marriages$Continent), mean)

Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], 

Group.1       Country Continent MarriageRate DivorceRate Ratio    Percent 
1               NA      NA         7.233333    2.633333    2.740000 39.56000
2 Africa        NA      NA        10.166667    1.650000    6.211667 16.89833
3 Asia          NA      NA         7.124000    1.688000    4.823600 26.15360
4 Europe        NA      NA         5.237209    1.962791    3.134651 38.79791
5 North America NA      NA         5.000000    1.512500    3.944375 30.30375
6 Oceania       NA      NA         5.500000    1.733333    3.540000 34.01667
7 South America NA      NA         4.216667    1.066667    3.963333 26.63333

In [85]:
continents_summary

marriages$Continent: 
[1] 39,56
------------------------------------------------------------ 
marriages$Continent: Africa
[1] 16,89833
------------------------------------------------------------ 
marriages$Continent: Asia
[1] 26,1536
------------------------------------------------------------ 
marriages$Continent: Europe
[1] 38,79791
------------------------------------------------------------ 
marriages$Continent: North America
[1] 30,30375
------------------------------------------------------------ 
marriages$Continent: Oceania
[1] 34,01667
------------------------------------------------------------ 
marriages$Continent: South America
[1] 26,63333

In [87]:
max(marriages$MarriageRate)
min(marriages$MarriageRate)

[1] 17,4

[1] 2,3

In [90]:
cut_points <- c(0, 5,10, 20)

cut(marriages$MarriageRate, cut_points)

[1] (5,10]  (10,20] (5,10]  (0,5]   (5,10]  (5,10]  (5,10]  (5,10]  (0,5]  
 [10] (10,20] (0,5]   (5,10]  (0,5]   (0,5]   (0,5]   (5,10]  (0,5]   (5,10] 
 [19] (0,5]   (5,10]  (5,10]  (5,10]  (5,10]  (0,5]   (5,10]  (10,20] (0,5]  
 [28] (0,5]   (0,5]   (0,5]   (0,5]   (5,10]  (0,5]   (5,10]  (0,5]   (0,5]  
 [37] (0,5]   (5,10]  (0,5]   (10,20] (0,5]   (5,10]  (0,5]   (5,10]  (0,5]  
 [46] (10,20] (5,10]  (5,10]  (5,10]  (5,10]  (5,10]  (10,20] (0,5]   (5,10] 
 [55] (0,5]   (5,10]  (5,10]  (5,10]  (5,10]  (5,10]  (0,5]   (5,10]  (0,5]  
 [64] (0,5]   (0,5]   (5,10]  (0,5]   (0,5]   (5,10]  (0,5]   (0,5]   (5,10] 
 [73] (5,10]  (0,5]   (5,10]  (5,10]  (5,10]  (5,10]  (10,20] (5,10]  (5,10] 
 [82] (0,5]   (0,5]   (0,5]   (0,5]   (0,5]   (0,5]   (0,5]   (10,20] (5,10] 
 [91] (10,20] (5,10]  (5,10]  (5,10]  (5,10]  (5,10]  (0,5]   (0,5]   (5,10] 
[100] (5,10]  (0,5]   (5,10] 
Levels: (0,5] (5,10] (10,20]

In [91]:
marriage_rates_nominal <- cut(marriages$MarriageRate, cut_points)
levels(marriage_rates_nominal) <- c("low", "medium", "high")

marriage_rates_nominal

[1] medium high   medium low    medium medium medium medium low    high  
 [11] low    medium low    low    low    medium low    medium low    medium
 [21] medium medium medium low    medium high   low    low    low    low   
 [31] low    medium low    medium low    low    low    medium low    high  
 [41] low    medium low    medium low    high   medium medium medium medium
 [51] medium high   low    medium low    medium medium medium medium medium
 [61] low    medium low    low    low    medium low    low    medium low   
 [71] low    medium medium low    medium medium medium medium high   medium
 [81] medium low    low    low    low    low    low    low    high   medium
 [91] high   medium medium medium medium medium low    low    medium medium
[101] low    medium
Levels: low medium high

In [92]:
marriages$MarriageRateNominal <- marriage_rates_nominal

marriages

Country                          Continent     MarriageRate DivorceRate
1   Albania                          Europe         8,1         1,7        
2   Algeria                          Africa        10,1         1,6        
3   Armenia                          Europe         6,0         1,0        
4   Australia (more info)            Oceania        4,6         2,0        
5   Austria                          Europe         5,3         1,8        
6   Azerbaijan                       Asia           9,7         1,5        
7   Bahamas                          North America  6,1         1,0        
8   Belarus                          Europe         9,2         4,1        
9   Belgium (more info)              Europe         3,9         2,1        
10  Bermuda                                        10,6         2,7        
11  Bosnia and Herzegovina           Europe         4,8         0,6        
12  Brazil                           South America  6,6         1,4        
13  Bulgaria                         Europe         4,1         1,5        
14  Canada (more info)               North America  4,4         2,1        
15  Chile                            South America  3,3         0,7        
16  China                            Asia           7,2         3,2        
17  Colombia                         South America  2,3         0,7        
18  Costa Rica                       North America  5,3         2,5        
19  Croatia                          Europe         4,9         1,5        
20  Cuba                             North America  5,2         2,9        
21  Cyprus                           Asia           6,8         2,2        
22  Czech Republic                   Europe         5,1         2,3        
23  Denmark                          Europe         5,6         2,6        
24  Dominican Republic               North America  4,4         1,8        
25  Ecuador                          South America  5,6         1,1        
26  Egypt                            Africa        11,0         1,9        
27  El Salvador                      North America  3,5         0,8        
28  Estonia                          Europe         5,0         2,4        
29  EU                                              4,4         2,0        
30  Finland                          Europe         4,3         2,4        
⋮   ⋮                                ⋮             ⋮            ⋮          
73  Russia                           Europe         5,3         3,9        
74  Saint Lucia                      North America  2,8         0,7        
75  Saint Vincent and the Grenadines North America  5,8         0,8        
76  San Marino                       Europe         6,1         2,5        
77  Saudi Arabia                     Asia           5,6         2,1        
78  Serbia                           Europe         5,2         1,4        
79  Seychelles                       Africa        17,4         1,9        
80  Singapore                        Asia           6,8         1,9        
81  Slovakia                         Europe         5,7         1,8        
82  Slovenia                         Europe         3,5         1,1        
83  South Africa (more info)         Africa         3,5         0,6        
84  South Korea (more info)          Asia           4,7         2,2        
85  Spain                            Europe         3,5         2,0        
87  Suriname                         South America  4,2         1,3        
88  Sweden                           Europe         5,0         2,5        
89  Switzerland                      Europe         4,8         2,0        
90  Syria                            Asia          10,6         1,3        
91  Taiwan                           Asia           5,7         2,3        
92  Tajikistan                       Asia          13,5         1,4        
93  Thailand                         Asia           5,5         1,4        
94  Tonga                            Oceania        

In [101]:
marriages[marriages$MarriageRateNominal == 'low',]

Country                    Continent     MarriageRate DivorceRate Ratio
4   Australia (more info)      Oceania       4,6          2,0         2,30 
9   Belgium (more info)        Europe        3,9          2,1         1,86 
11  Bosnia and Herzegovina     Europe        4,8          0,6         8,00 
13  Bulgaria                   Europe        4,1          1,5         2,73 
14  Canada (more info)         North America 4,4          2,1         2,10 
15  Chile                      South America 3,3          0,7         4,71 
17  Colombia                   South America 2,3          0,7         3,29 
19  Croatia                    Europe        4,9          1,5         3,27 
24  Dominican Republic         North America 4,4          1,8         2,44 
27  El Salvador                North America 3,5          0,8         4,38 
28  Estonia                    Europe        5,0          2,4         2,08 
29  EU                                       4,4          2,0         2,20 
30  Finland                    Europe        4,3          2,4         1,79 
31  France                     Europe        3,7          1,9         1,95 
33  Germany                    Europe        4,9          1,9         2,58 
35  Grenada                    North America 5,0          1,1         4,55 
36  Greece                     Europe        4,7          1,8         2,61 
37  Guatemala                  North America 3,8          1,2         3,17 
39  Iceland                    Europe        4,9          1,8         2,72 
41  Ireland                    Europe        4,6          0,7         6,57 
43  Italy                      Europe        3,2          1,5         2,13 
45  Japan (more info)          Asia          4,8          1,7         2,82 
53  Liechtenstein              Europe        5,0          2,4         2,08 
55  Luxembourg                 Europe        3,1          2,0         1,55 
61  Mongolia                   Asia          3,4          1,1         3,09 
63  Netherlands                Europe        3,7          1,8         2,06 
64  New Zealand (more info)    Oceania       4,8          2,0         2,40 
65  Nicaragua                  North America 4,5          0,8         5,63 
67  Norway (more info)         Europe        4,3          1,9         2,26 
68  Panama                     North America 3,7          1,0         3,70 
70  Portugal                   Europe        3,4          2,0         1,70 
71  Qatar                      Asia          3,3          1,1         3,00 
74  Saint Lucia                North America 2,8          0,7         4,00 
82  Slovenia                   Europe        3,5          1,1         3,18 
83  South Africa (more info)   Africa        3,5          0,6         5,83 
84  South Korea (more info)    Asia          4,7          2,2         2,14 
85  Spain                      Europe        3,5          2,0         1,75 
87  Suriname                   South America 4,2          1,3         3,23 
88  Sweden                     Europe        5,0          2,5         2,00 
89  Switzerland                Europe        4,8          2,0         2,40 
99  United Arab Emirates       Asia          2,8          0,7         4,00 
100 United Kingdom (more info) Europe        4,4          1,8         2,44 
104 Venezuela                  South America 3,3          1,2         2,75 
    Percent MarriageRateNominal
4   43,48   low                
9   53,85   low                
11  12,50   low                
13  36,59   low                
14  47,73   low                
15  21,21   low                
17  30,43   low                
19  30,61   low                
24  40,91   low                
27  22,86   low                
28  48,00   low                
29  45,45   low                
30  55,81   low                
31  51,35   low                
33  38,78   low                
35  22,00   low                
36  38,30   low                
37  31,58   low                
39  36,73   low                
41  15,22   low     

In [103]:
aggregate(
    marriages,
    list(
        marriages$Continent,
        marriages$MarriageRateNominal
    ),
    mean
)

Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], ...):
“argument is not numeric or logical: returning NA”Warning message in mean.default(X[[i]], 

Group.1       Group.2 Country Continent MarriageRate DivorceRate Ratio   
1                low     NA      NA         4,400000    2,000000    2,200000
2  Africa        low     NA      NA         3,500000    0,600000    5,830000
3  Asia          low     NA      NA         3,800000    1,360000    3,010000
4  Europe        low     NA      NA         4,259091    1,800000    2,714091
5  North America low     NA      NA         4,012500    1,187500    3,746250
6  Oceania       low     NA      NA         4,700000    2,000000    2,350000
7  South America low     NA      NA         3,275000    0,975000    3,495000
8                medium  NA      NA         6,700000    3,200000    2,090000
9  Africa        medium  NA      NA         8,200000    1,400000    5,860000
10 Asia          medium  NA      NA         7,100000    1,781250    4,801875
11 Europe        medium  NA      NA         6,261905    2,133333    3,575238
12 North America medium  NA      NA         5,987500    1,837500    4,142500
13 Oceania       medium  NA      NA         7,100000    1,200000    5,920000
14 South America medium  NA      NA         6,100000    1,250000    4,900000
15               high    NA      NA        10,600000    2,700000    3,930000
16 Africa        high    NA      NA        12,325000    1,975000    6,395000
17 Asia          high    NA      NA        11,375000    1,725000    7,177500
   Percent  MarriageRateNominal
1  45,45000 NA                 
2  17,14000 NA                 
3  34,58200 NA                 
4  43,17955 NA                 
5  29,36125 NA                 
6  42,57500 NA                 
7  29,73750 NA                 
8  47,76000 NA                 
9  17,07000 NA                 
10 26,15750 NA                 
11 34,20762 NA                 
12 31,24625 NA                 
13 16,90000 NA                 
14 20,42500 NA                 
15 25,47000 NA                 
16 16,79500 NA                 
17 15,60250 NA

In [23]:
religion <- read.table("religion.csv", header = T, sep = ",")
head(religion)

Rank Country    Important Unimportant
1 1    Estonia    16%       78%        
2 2    Sweden     17%       82%        
3 3    Denmark    19%       80%        
4 4    Czechia[a] 21%       75%        
5 5    Norway[a]  22%       78%        
6 6    Japan      24%       75%

In [26]:
?gsub

In [28]:
gsub("\\[a\\]","", "Czechia[a]")

[1] "Czechia"

In [37]:
strsplit("Czechia[a]", "\\[")[[1]][[1]]

[1] "Czechia"

In [38]:
sapply(
    religion$Country, 
    function(x) gsub("\\[a\\]","", x)
)

[1] "Estonia"                          "Sweden"                          
  [3] "Denmark"                          "Czechia"                         
  [5] "Norway"                           "Japan"                           
  [7] "Hong Kong"                        "United Kingdom"                  
  [9] "Finland"                          "Vietnam"                         
 [11] "France"                           "Australia"                       
 [13] "Netherlands"                      "New Zealand"                     
 [15] "Belgium"                          "Cuba"                            
 [17] "Bulgaria"                         "Russia"                          
 [19] "Belarus"                          "Luxembourg"                      
 [21] "Azerbaijan"                       "Hungary"                         
 [23] "Latvia"                           "Germany"                         
 [25] "Uruguay"                          "Switzerland"                     
 [27] "Canada"                           "Lithuania"                       
 [29] "South Korea"                      "Taiwan"                          
 [31] "Ukraine"                          "Slovenia"                        
 [33] "Slovakia"                         "Spain"                           
 [35] "Albania"                          "Israel"                          
 [37] "Kazakhstan"                       "Serbia"                          
 [39] "Ireland"                          "Austria"                         
 [41] "Uzbekistan"                       "Belize"                          
 [43] "Argentina"                        "United States"                   
 [45] "Nicaragua"                        "Croatia"                         
 [47] "Chile"                            "Singapore"                       
 [49] "Jamaica"                          "Montenegro"                      
 [51] "Greece"                           "Portugal"                        
 [53] "Italy"                            "Moldova"                         
 [55] "Kyrgyzstan"                       "Iran"                            
 [57] "Mexico"                           "Poland"                          
 [59] "Cyprus"                           "North Macedonia"                 
 [61] "Botswana"                         "Bosnia and Herzegovina"          
 [63] "Venezuela"                        "Costa Rica"                      
 [65] "Armenia"                          "Turkmenistan"                    
 [67] "Togo"                             "Georgia"                         
 [69] "Turkey"                           "Ecuador"                         
 [71] "Colombia"                         "Peru"                            
 [73] "Iraq"                             "Romania"                         
 [75] "South Africa"                     "Tajikistan"                      
 [77] "Haiti"                            "Mozambique"                      
 [79] "Malta"                            "Brazil"                          
 [81] "Dominican Republic"               "Lebanon"                         
 [83] "Zimbabwe"                         "Ivory Coast"                     
 [85] "Burkina Faso"                     "Panama"                          
 [87] "Angola"                           "Guatemala"                       
 [89] "Tanzania"                         "Bolivia"                         
 [91] "Syria"                            "India"                           
 [93] "United Arab Emirates"             "Kuwait"                          
 [95] "Namibia"                          "Trinidad and Tobago"             
 [97] "Paraguay"                         "Pakistan"                        
 [99] "State of Palestine"               "Sudan"                           
[101] "Uganda"                           "Madagascar"                      
[103] "Benin"                            "Nepal"                           
[105] "Tunisia"                          "Saudi Ar

In [39]:
religion$Country <- sapply(
    religion$Country, 
    function(x) gsub("\\[a\\]","", x)
)

head(religion)

Rank Country Important Unimportant
1 1    Estonia 16%       78%        
2 2    Sweden  17%       82%        
3 3    Denmark 19%       80%        
4 4    Czechia 21%       75%        
5 5    Norway  22%       78%        
6 6    Japan   24%       75%

In [40]:
sapply(religion, class)

Rank     Country   Important Unimportant 
  "integer" "character"    "factor"    "factor"

In [41]:
as.vector(religion$Important)

[1] "16%"  "17%"  "19%"  "21%"  "22%"  "24%"  "24%"  "27%"  "28%"  "30%" 
 [11] "30%"  "32%"  "33%"  "33%"  "33%"  "34%"  "34%"  "34%"  "34%"  "39%" 
 [21] "39%"  "39%"  "39%"  "40%"  "41%"  "41%"  "42%"  "42%"  "43%"  "45%" 
 [31] "46%"  "47%"  "47%"  "49%"  "50%"  "51%"  "53%"  "54%"  "54%"  "55%" 
 [41] "59%"  "62%"  "65%"  "69%"  "70%"  "70%"  "70%"  "70%"  "70%"  "71%" 
 [51] "71%"  "72%"  "72%"  "72%"  "72%"  "73%"  "73%"  "75%"  "75%"  "76%" 
 [61] "77%"  "77%"  "79%"  "79%"  "79%"  "80%"  "80%"  "81%"  "82%"  "82%" 
 [71] "83%"  "84%"  "84%"  "84%"  "85%"  "85%"  "85%"  "86%"  "86%"  "87%" 
 [81] "87%"  "87%"  "88%"  "88%"  "88%"  "88%"  "88%"  "88%"  "89%"  "89%" 
 [91] "89%"  "90%"  "91%"  "91%"  "92%"  "92%"  "92%"  "92%"  "93%"  "93%" 
[101] "93%"  "93%"  "93%"  "93%"  "93%"  "93%"  "94%"  "94%"  "94%"  "94%" 
[111] "94%"  "95%"  "95%"  "95%"  "95%"  "95%"  "95%"  "95%"  "95%"  "96%" 
[121] "96%"  "96%"  "96%"  "96%"  "96%"  "96%"  "97%"  "97%"  "97%"  "97%" 
[131] "97%"  "97%"  "97%"  "97%"  "98%"  "98%"  "98%"  "99%"  "99%"  "99%" 
[141] "100%" "100%" "100%" "100%" "100%"

In [43]:
sapply(
    religion$Important, 
    function(x) gsub("%","", x)
)

[1] "16"  "17"  "19"  "21"  "22"  "24"  "24"  "27"  "28"  "30"  "30"  "32" 
 [13] "33"  "33"  "33"  "34"  "34"  "34"  "34"  "39"  "39"  "39"  "39"  "40" 
 [25] "41"  "41"  "42"  "42"  "43"  "45"  "46"  "47"  "47"  "49"  "50"  "51" 
 [37] "53"  "54"  "54"  "55"  "59"  "62"  "65"  "69"  "70"  "70"  "70"  "70" 
 [49] "70"  "71"  "71"  "72"  "72"  "72"  "72"  "73"  "73"  "75"  "75"  "76" 
 [61] "77"  "77"  "79"  "79"  "79"  "80"  "80"  "81"  "82"  "82"  "83"  "84" 
 [73] "84"  "84"  "85"  "85"  "85"  "86"  "86"  "87"  "87"  "87"  "88"  "88" 
 [85] "88"  "88"  "88"  "88"  "89"  "89"  "89"  "90"  "91"  "91"  "92"  "92" 
 [97] "92"  "92"  "93"  "93"  "93"  "93"  "93"  "93"  "93"  "93"  "94"  "94" 
[109] "94"  "94"  "94"  "95"  "95"  "95"  "95"  "95"  "95"  "95"  "95"  "96" 
[121] "96"  "96"  "96"  "96"  "96"  "96"  "97"  "97"  "97"  "97"  "97"  "97" 
[133] "97"  "97"  "98"  "98"  "98"  "99"  "99"  "99"  "100" "100" "100" "100"
[145] "100"

In [44]:
convert_percent <- function(vec) {
    sapply(
        vec, 
        function(x) gsub("%","", x)
    )
}

convert_percent(religion$Important)
convert_percent(religion$Unimportant)

[1] "16"  "17"  "19"  "21"  "22"  "24"  "24"  "27"  "28"  "30"  "30"  "32" 
 [13] "33"  "33"  "33"  "34"  "34"  "34"  "34"  "39"  "39"  "39"  "39"  "40" 
 [25] "41"  "41"  "42"  "42"  "43"  "45"  "46"  "47"  "47"  "49"  "50"  "51" 
 [37] "53"  "54"  "54"  "55"  "59"  "62"  "65"  "69"  "70"  "70"  "70"  "70" 
 [49] "70"  "71"  "71"  "72"  "72"  "72"  "72"  "73"  "73"  "75"  "75"  "76" 
 [61] "77"  "77"  "79"  "79"  "79"  "80"  "80"  "81"  "82"  "82"  "83"  "84" 
 [73] "84"  "84"  "85"  "85"  "85"  "86"  "86"  "87"  "87"  "87"  "88"  "88" 
 [85] "88"  "88"  "88"  "88"  "89"  "89"  "89"  "90"  "91"  "91"  "92"  "92" 
 [97] "92"  "92"  "93"  "93"  "93"  "93"  "93"  "93"  "93"  "93"  "94"  "94" 
[109] "94"  "94"  "94"  "95"  "95"  "95"  "95"  "95"  "95"  "95"  "95"  "96" 
[121] "96"  "96"  "96"  "96"  "96"  "96"  "97"  "97"  "97"  "97"  "97"  "97" 
[133] "97"  "97"  "98"  "98"  "98"  "99"  "99"  "99"  "100" "100" "100" "100"
[145] "100"

[1] "78" "82" "80" "75" "78" "75" "74" "73" "70" "69" "69" "68" "67" "66" "58"
 [16] "64" "62" "66" "56" "59" "53" "58" "58" "59" "59" "57" "57" "49" "56" "54"
 [31] "48" "52" "52" "51" "49" "48" "48" "44" "46" "43" "38" "33" "34" "31" "30"
 [46] "28" "29" "29" "30" "28" "28" "26" "25" "19" "25" "27" "25" "19" "25" "22"
 [61] "23" "21" "21" "20" "19" "18" "13" "16" "15" "17" "16" "14" "11" "12" "15"
 [76] "12" "12" "14" "10" "13" "13" "12" "12" "12" "12" "11" "11" "9"  "11" "10"
 [91] "9"  "10" "8"  "6"  "9"  "8"  "8"  "8"  "7"  "7"  "7"  "7"  "7"  "6"  "5" 
[106] "4"  "6"  "6"  "6"  "5"  "4"  "5"  "5"  "4"  "4"  "5"  "5"  "5"  "3"  "4" 
[121] "4"  "3"  "3"  "3"  "4"  "4"  "3"  "3"  "3"  "3"  "2"  "2"  "2"  "2"  "2" 
[136] "2"  "2"  "1"  "1"  "1"  "0"  "0"  "0"  "0"  "0"

In [47]:
religion$Important <- as.numeric(convert_percent(religion$Important))
religion$Unimportant <- as.numeric(convert_percent(religion$Unimportant))

head(religion)

Rank Country Important Unimportant
1 1    Estonia 16        78         
2 2    Sweden  17        82         
3 3    Denmark 19        80         
4 4    Czechia 21        75         
5 5    Norway  22        78         
6 6    Japan   24        75

In [48]:
religion$Ratio <- religion$Important / religion$Unimportant

head(religion)

Rank Country Important Unimportant Ratio    
1 1    Estonia 16        78          0.2051282
2 2    Sweden  17        82          0.2073171
3 3    Denmark 19        80          0.2375000
4 4    Czechia 21        75          0.2800000
5 5    Norway  22        78          0.2820513
6 6    Japan   24        75          0.3200000

In [52]:
head(religion)
head(marriages)

Rank Country Important Unimportant Ratio    
1 1    Estonia 16        78          0,2051282
2 2    Sweden  17        82          0,2073171
3 3    Denmark 19        80          0,2375000
4 4    Czechia 21        75          0,2800000
5 5    Norway  22        78          0,2820513
6 6    Japan   24        75          0,3200000

Country               Continent MarriageRate DivorceRate Ratio Percent
1 Albania               Europe     8,1         1,7         4,76  20,99  
2 Algeria               Africa    10,1         1,6         6,31  15,84  
3 Armenia               Europe     6,0         1,0         6,00  16,67  
4 Australia (more info) Oceania    4,6         2,0         2,30  43,48  
5 Austria               Europe     5,3         1,8         2,94  33,96  
6 Azerbaijan            Asia       9,7         1,5         6,47  15,46

## Extramarital Affairs

https://fairmodel.econ.yale.edu/rayfair/pdf/1978a200.pdf

"The  purpose  of  this paper  is to consider  the determinants of  leisure  time 
spent in one particular type of activity with non-household  members: 
extramarital  affairs.  The extent  of  such activity is by  no  means  small.  In 
one  of  two  surveys  used for  the  empirical work  in this study,  27.2  per- 
cent  of  the  first-time-married working men  and  22.9  percent  of the  first- 
time-married working women were  having an  extramarital affair  at  the 
time  of  the  survey."

https://vincentarelbundock.github.io/Rdatasets/doc/COUNT/affairs.html

Usage

data(affairs)

Format

A data frame with 601 observations on the following 18 variables.

naffairs

    number of affairs within last year
kids

    1=have children;0= no children
vryunhap

    (1/0) very unhappily married
unhap

    (1/0) unhappily married
avgmarr

    (1/0) average married
hapavg

    (1/0) happily married
vryhap

    (1/0) very happily married
antirel

    (1/0) anti religious
notrel

    (1/0) not religious
slghtrel

    (1/0) slightly religious
smerel

    (1/0) somewhat religious
vryrel

    (1/0) very religious
yrsmarr1

    (1/0) >0.75 yrs
yrsmarr2

    (1/0) >1.5 yrs
yrsmarr3

    (1/0) >4.0 yrs
yrsmarr4

    (1/0) >7.0 yrs
yrsmarr5

    (1/0) >10.0 yrs
yrsmarr6

    (1/0) >15.0 yrs

In [ ]:
affairs_df <- read.table("affairs.csv", header = TRUE, sep = ",")
head(affairs_df)

In [ ]:
tail(affairs_df)

In [ ]:
cont_table <- table(
    affairs_df[c("naffairs", "kids")]
    , dnn = c("Affairs", "Kids (yes/no)")
)
cont_table

In [ ]:
class(cont_table)

In [ ]:
matrix(
    cont_table
    , ncol = ncol(cont_table)
    , nrow = nrow(cont_table)
    , dimnames = dimnames(cont_table)
)

In [ ]:
counts <- as.data.frame.matrix(
    cont_table
)
colnames(counts) <- c("Kids no", "Kids yes")
counts

In [ ]:
ret <- c()
for (i in 1:nrow(counts)) {
    row <- counts[i,]
    ret <- c(ret, row["Kids yes"] / row["Kids no"])
}
ret

In [ ]:
apply(counts, 1, function(row) {
        row["Kids yes"] / row["Kids no"]
    }
)

In [ ]:
apply(counts, 2, function(x) {sum(x)})

In [ ]:
sum(counts$"Kids yes")

In [ ]:
l_x <- list(
    prvni = seq(1, 10, 0.5)
    , druha = matrix(11:20, ncol = 2)
    , treti = as.character(1:10)
)
l_x

In [ ]:
object_info <- function(x) {
    list(
        trida = class(x)
        , typ = typeof(x)
        , mode = mode(x)
    )
}

info <- lapply(l_x, object_info)
info[["prvni"]]
info[["prvni"]][["mode"]]

In [ ]:
unlist(lapply(1:10, function(x) {x / 10}))

In [ ]:
sapply(1:10, function(x) {x / 10}, simplify = F)

In [ ]:
lapply(affairs_df, function(x) {sum(x) / length(x)})

In [ ]:
head(affairs_df)

In [ ]:
aggregate_vals <- function(row) {
    return(which(row == 1))
}

cols_to_levels <- function(data, cols) {
    data_subset <- data[cols]
    return(
        apply(data_subset, 1, aggregate_vals)
    )
}

happy_cols <- c("vryunhap", "unhap", "avgmarr", "hapavg", "vryhap")
affairs_df["Happiness"] <- cols_to_levels(
    affairs_df
    , happy_cols
)

In [ ]:
head(affairs_df)

In [ ]:
affairs_df[happy_cols] <- NULL
head(affairs_df)

In [ ]:
years_cols <- c("yrsmarr1", "yrsmarr2", "yrsmarr3", "yrsmarr4", "yrsmarr5", "yrsmarr6")
affairs_df["Years"] <- cols_to_levels(
    affairs_df
    , years_cols
)
affairs_df[years_cols] <- NULL
head(affairs_df)

In [ ]:
rel_cols <- c("antirel", "notrel", "slghtrel", "smerel", "vryrel")
affairs_df["Religion"] <- cols_to_levels(
    affairs_df
    , rel_cols
)
affairs_df[rel_cols] <- NULL
head(affairs_df)

In [ ]:
rel_happ <- table(
    affairs_df[["Religion"]]
    , affairs_df[["Happiness"]]
    , dnn = c("Religion Level", "Happiness Level")
)
rel_happ

In [ ]:
rel_happ <- round(rel_happ / nrow(affairs_df), digits = 2)
rel_happ

In [ ]:
image(1:ncol(rel_happ), 1:nrow(rel_happ), t(rel_happ), col = heat.colors(256), axes = FALSE)
axis(1, 1:ncol(rel_happ), colnames(rel_happ))
axis(2, 1:nrow(rel_happ), rownames(rel_happ))
for (x in 1:ncol(rel_happ))
  for (y in 1:nrow(rel_happ))
    text(x, y, rel_happ[y,x])

In [ ]:
head(affairs_df)

In [ ]:
output <- by(
    affairs_df
    , affairs_df[["Years"]]
    , function(x) {
        list(
            kid_ratio = sum(x[["kids"]]) / nrow(x)
            , affairs_ratio = sum(x[["naffairs"]] > 0) / nrow(x)
        )
    }
)
output

In [ ]:
do.call(rbind, output)

In [ ]:
rbind(1:10, 11:20)

In [ ]:
cbind(1:10, 11:20)

In [ ]:
x1 <- c(1, 2, 5)
x2 <- c(3, 4, 8)

do.call(sum, list(x1, x2))
do.call(rbind, list(x1, x2))

V podstatě stejné jako:

In [ ]:
sum(x1, x2)
rbind(x1, x2)